From ed201f4a5169e58c43611c429a593d52be7d31d7 Mon Sep 17 00:00:00 2001
From: Allen Ye <allen.ye@mediatek.com>
Date: Thu, 19 Dec 2024 19:54:34 +0800
Subject: [PATCH 048/119] mtk: mt76: mt7996: Change beacon protection fw cmd

Change beacon protection fw cmd from BSS_INFO to STA_REC_UPDATE
and fix memory leakage in mt7996_mcu_add_key

Signed-off-by: Allen Ye <allen.ye@mediatek.com>

Remove mt7996_mcu_get_pn().
When installing BIGTK, always provide PN=0 to FW.
This avoids the situation where PN is not synchronized between AP
and STA after the AP rekeys

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 mt7996/main.c   |  10 ++---
 mt7996/mcu.c    | 114 ++++++++++--------------------------------------
 mt7996/mcu.h    |  14 ++----
 mt7996/mt7996.h |   4 --
 4 files changed, 31 insertions(+), 111 deletions(-)

diff --git a/mt7996/main.c b/mt7996/main.c
index 114ef789..e569365f 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -705,6 +705,7 @@ static int mt7996_set_key(struct ieee80211_hw *hw, enum set_key_cmd cmd,
 			return -EOPNOTSUPP;
 		}
 
+		/* Todo: remove me after fix set dtim period to fw */
 		if (cmd == SET_KEY && !sta && !mconf->mt76.cipher) {
 			mconf->mt76.cipher = mt76_connac_mcu_get_cipher(key->cipher);
 			mt7996_mcu_add_bss_info(mconf->phy, vif, conf, &mconf->mt76, msta_link, true);
@@ -720,12 +721,9 @@ static int mt7996_set_key(struct ieee80211_hw *hw, enum set_key_cmd cmd,
 
 		mt76_wcid_key_setup(&dev->mt76, &msta_link->wcid, key);
 
-		if (key->keyidx == 6 || key->keyidx == 7)
-			err = mt7996_mcu_bcn_prot_enable(dev, mconf, msta_link, key);
-		else
-			err = mt7996_mcu_add_key(&dev->mt76, mconf, key,
-						 MCU_WMWA_UNI_CMD(STA_REC_UPDATE),
-						 &msta_link->wcid, cmd);
+		err = mt7996_mcu_add_key(&dev->mt76, mconf, key,
+					 MCU_WMWA_UNI_CMD(STA_REC_UPDATE),
+					 &msta_link->wcid, cmd);
 	}
 out:
 	mutex_unlock(&dev->mt76.mutex);
diff --git a/mt7996/mcu.c b/mt7996/mcu.c
index bf52c8b4..13123c36 100644
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -3657,7 +3657,8 @@ int mt7996_mcu_teardown_mld_sta(struct mt7996_dev *dev,
 }
 
 static int
-mt7996_mcu_sta_key_tlv(struct mt76_wcid *wcid,
+mt7996_mcu_sta_key_tlv(struct mt76_dev *dev,
+		       struct mt76_wcid *wcid,
 		       struct sk_buff *skb,
 		       struct ieee80211_key_conf *key,
 		       enum set_key_cmd cmd)
@@ -3686,6 +3687,25 @@ mt7996_mcu_sta_key_tlv(struct mt76_wcid *wcid,
 		sec_key->key_len = key->keylen;
 		sec_key->need_resp = 0;
 		memcpy(sec_key->key, key->key, key->keylen);
+		/* Todo: move the BCN protection logic into mt76_connac_mcu_get_cipher() */
+		if (sec_key->key_id == 6 || sec_key->key_id == 7) {
+			switch (key->cipher) {
+			case WLAN_CIPHER_SUITE_AES_CMAC:
+				sec_key->cipher_id = MCU_CIPHER_BCN_PROT_CMAC_128;
+				break;
+			case WLAN_CIPHER_SUITE_BIP_GMAC_128:
+				sec_key->cipher_id = MCU_CIPHER_BCN_PROT_GMAC_128;
+				break;
+			case WLAN_CIPHER_SUITE_BIP_GMAC_256:
+				sec_key->cipher_id = MCU_CIPHER_BCN_PROT_GMAC_256;
+				break;
+			case WLAN_CIPHER_SUITE_BIP_CMAC_256:
+			default:
+				dev_err(dev->dev, "Unsupported BIGTK cipher\n");
+				return -EOPNOTSUPP;
+			}
+			sec_key->bcn_mode = BP_SW_MODE;
+		}
 
 		if (cipher == MCU_CIPHER_TKIP) {
 			/* Rx/Tx MIC keys are swapped */
@@ -3713,101 +3733,15 @@ int mt7996_mcu_add_key(struct mt76_dev *dev, struct mt7996_vif_link *mconf,
 	if (IS_ERR(skb))
 		return PTR_ERR(skb);
 
-	ret = mt7996_mcu_sta_key_tlv(wcid, skb, key, cmd);
-	if (ret)
-		return ret;
-
-	return mt76_mcu_skb_send_msg(dev, skb, mcu_cmd, true);
-}
-
-static int mt7996_mcu_get_pn(struct mt7996_dev *dev,
-			     struct mt7996_vif_link *link,
-			     struct mt7996_sta_link *msta_link, u8 *pn)
-{
-#define TSC_TYPE_BIGTK_PN 2
-	struct sta_rec_pn_info *pn_info;
-	struct sk_buff *skb, *rskb;
-	struct tlv *tlv;
-	int ret;
-
-	skb = mt76_connac_mcu_alloc_sta_req(&dev->mt76, &link->mt76,
-					    &msta_link->wcid);
-	if (IS_ERR(skb))
-		return PTR_ERR(skb);
-
-	tlv = mt76_connac_mcu_add_tlv(skb, STA_REC_PN_INFO, sizeof(*pn_info));
-	pn_info = (struct sta_rec_pn_info *)tlv;
-
-	pn_info->tsc_type = TSC_TYPE_BIGTK_PN;
-	ret = mt76_mcu_skb_send_and_get_msg(&dev->mt76, skb,
-					    MCU_WM_UNI_CMD_QUERY(STA_REC_UPDATE),
-					    true, &rskb);
-	if (ret)
-		return ret;
-
-	skb_pull(rskb, 4);
-
-	pn_info = (struct sta_rec_pn_info *)rskb->data;
-	if (le16_to_cpu(pn_info->tag) == STA_REC_PN_INFO)
-		memcpy(pn, pn_info->pn, 6);
-
-	dev_kfree_skb(rskb);
-	return 0;
-}
-
-int mt7996_mcu_bcn_prot_enable(struct mt7996_dev *dev,
-			       struct mt7996_vif_link *link,
-			       struct mt7996_sta_link *msta_link,
-			       struct ieee80211_key_conf *key)
-{
-	struct mt7996_mcu_bcn_prot_tlv *bcn_prot;
-	struct sk_buff *skb;
-	struct tlv *tlv;
-	u8 pn[6] = {};
-	int len = sizeof(struct bss_req_hdr) +
-		  sizeof(struct mt7996_mcu_bcn_prot_tlv);
-	int ret;
-
-	skb = __mt7996_mcu_alloc_bss_req(&dev->mt76, &link->mt76, len);
-	if (IS_ERR(skb))
-		return PTR_ERR(skb);
-
-	tlv = mt76_connac_mcu_add_tlv(skb, UNI_BSS_INFO_BCN_PROT, sizeof(*bcn_prot));
-
-	bcn_prot = (struct mt7996_mcu_bcn_prot_tlv *)tlv;
-
-	ret = mt7996_mcu_get_pn(dev, link, msta_link, pn);
+	ret = mt7996_mcu_sta_key_tlv(dev, wcid, skb, key, cmd);
 	if (ret) {
 		dev_kfree_skb(skb);
 		return ret;
 	}
 
-	switch (key->cipher) {
-	case WLAN_CIPHER_SUITE_AES_CMAC:
-		bcn_prot->cipher_id = MCU_CIPHER_BCN_PROT_CMAC_128;
-		break;
-	case WLAN_CIPHER_SUITE_BIP_GMAC_128:
-		bcn_prot->cipher_id = MCU_CIPHER_BCN_PROT_GMAC_128;
-		break;
-	case WLAN_CIPHER_SUITE_BIP_GMAC_256:
-		bcn_prot->cipher_id = MCU_CIPHER_BCN_PROT_GMAC_256;
-		break;
-	case WLAN_CIPHER_SUITE_BIP_CMAC_256:
-	default:
-		dev_err(dev->mt76.dev, "Not supported Bigtk Cipher\n");
-		dev_kfree_skb(skb);
-		return -EOPNOTSUPP;
-	}
-
-	pn[0]++;
-	memcpy(bcn_prot->pn, pn, 6);
-	bcn_prot->enable = BP_SW_MODE;
-	memcpy(bcn_prot->key, key->key, WLAN_MAX_KEY_LEN);
-	bcn_prot->key_id = key->keyidx;
-
-	return mt76_mcu_skb_send_msg(&dev->mt76, skb,
-				     MCU_WMWA_UNI_CMD(BSS_INFO_UPDATE), true);
+	return mt76_mcu_skb_send_msg(dev, skb, mcu_cmd, true);
 }
+
 int mt7996_mcu_add_dev_info(struct mt7996_phy *phy, struct ieee80211_vif *vif,
 			    struct ieee80211_bss_conf *link_conf,
 			    struct mt76_vif_link *mlink, bool enable)
diff --git a/mt7996/mcu.h b/mt7996/mcu.h
index f44fabe0..d7975a07 100644
--- a/mt7996/mcu.h
+++ b/mt7996/mcu.h
@@ -491,17 +491,6 @@ enum {
 	BP_HW_MODE,
 };
 
-struct mt7996_mcu_bcn_prot_tlv {
-	__le16 tag;
-	__le16 len;
-	u8 pn[6];
-	u8 enable;
-	u8 cipher_id;
-	u8 key[WLAN_MAX_KEY_LEN];
-	u8 key_id;
-	u8 __rsv[3];
-} __packed;
-
 struct bss_ra_tlv {
 	__le16 tag;
 	__le16 len;
@@ -735,6 +724,9 @@ struct sec_key_uni {
 	u8 key_len;
 	u8 need_resp;
 	u8 key[32];
+	u8 pn[6];
+	u8 bcn_mode;
+	u8 _rsv;
 } __packed;
 
 struct sta_rec_sec_uni {
diff --git a/mt7996/mt7996.h b/mt7996/mt7996.h
index abcfbe91..a3312596 100644
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -1372,10 +1372,6 @@ bool mt7996_debugfs_rx_log(struct mt7996_dev *dev, const void *data, int len);
 int mt7996_mcu_add_key(struct mt76_dev *dev, struct mt7996_vif_link *mconf,
 		       struct ieee80211_key_conf *key, int mcu_cmd,
 		       struct mt76_wcid *wcid, enum set_key_cmd cmd);
-int mt7996_mcu_bcn_prot_enable(struct mt7996_dev *dev,
-			       struct mt7996_vif_link *link,
-			       struct mt7996_sta_link *msta_link,
-			       struct ieee80211_key_conf *key);
 int mt7996_mcu_wtbl_update_hdr_trans(struct mt7996_dev *dev,
 				     struct ieee80211_vif *vif,
 				     struct mt7996_vif_link *link,
-- 
2.45.2

